### PR TITLE
Add --filter to podman system prune

### PR DESCRIPTION
**Pull request by rhatdan**
Also document the allowable filters in podman system prune, podman image prune
and podman container prune.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
Fixes: https://github.com/containers/podman/issues/8453

**Comment by rhatdan** [comment]
@svdHero PTAL

**Comment by rhatdan** [comment]
@bblenard  PTAL

**Comment by TomSweeneyRedHat** [comment]
LGTM

**Comment by bblenard** [comment]
What about filters on volume pruning, or was that not part of this feature request?

**Comment by rhatdan** [comment]
I think volumes have to be handled separately,

podman volume prune itself does not currently support filters.

Another thing we need to add is the ability to prune networks, now that we have network support. but I think those need to be done separatly.  Lets get this one in for now.

**Comment by rhatdan** [comment]
@containers/podman-maintainers PTAL
@bblenard I opened issues on volume --filter and on podman network prune, if you want to take a look at those.

File: docs/source/markdown/podman-container-prune.1.md [hunk 1]
```diff
@@ -15,6 +15,17 @@ podman-container-prune - Remove all stopped containers from local storage
...
 Provide filter values.
 
+The --filter flag format is of “key=value”. If there is more than one filter, then pass multiple flags (e.g., --filter "foo=bar" --filter "bif=baz")
+
+Supported filters:
+
+- `until` (_timestamp_) - only remove containers and images created before given timestamp
+- `label` (label=_key_, label=_key=value_, label!=_key_, or label!=_key=value_) - only remove containers and images, with (or without, in case label!=... is used) the specified labels.
+
+The until filter can be Unix timestamps, date formatted timestamps, or Go duration strings (e.g. 10m, 1h30m) computed relative to the daemon machine’s time.
```
Comment (ashley-cui) [inline docs/source/markdown/podman-container-prune.1.md:25]: No daemons! 

```suggestion
The until filter can be Unix timestamps, date formatted timestamps, or Go duration strings (e.g. 10m, 1h30m) computed relative to the machine’s time.
```

File: docs/source/markdown/podman-image-prune.1.md [hunk 2]
```diff
@@ -22,6 +22,17 @@ Remove dangling images and images that have no associated containers.
...
 Provide filter values.
 
+The --filter flag format is of “key=value”. If there is more than one filter, then pass multiple flags (e.g., --filter "foo=bar" --filter "bif=baz")
+
+Supported filters:
+
+- `until` (_timestamp_) - only remove containers and images created before given timestamp
+- `label` (label=_key_, label=_key=value_, label!=_key_, or label!=_key=value_) - only remove containers and images, with (or without, in case label!=... is used) the specified labels.
+
+The until filter can be Unix timestamps, date formatted timestamps, or Go duration strings (e.g. 10m, 1h30m) computed relative to the daemon machine’s time.
```
Comment (ashley-cui) [inline docs/source/markdown/podman-image-prune.1.md:32]: Same as before
```suggestion
The until filter can be Unix timestamps, date formatted timestamps, or Go duration strings (e.g. 10m, 1h30m) computed relative to the machine’s time. 
```

File: docs/source/markdown/podman-system-prune.1.md [hunk 3]
```diff
@@ -18,6 +18,21 @@ By default, volumes are not removed to prevent important data from being deleted
...
+Provide filter values.
+
+The --filter flag format is of “key=value”. If there is more than one filter, then pass multiple flags (e.g., --filter "foo=bar" --filter "bif=baz")
+
+Supported filters:
+
+- `until` (_timestamp_) - only remove containers and images created before given timestamp
+- `label` (label=_key_, label=_key=value_, label!=_key_, or label!=_key=value_) - only remove containers and images, with (or without, in case label!=... is used) the specified labels.
+
+The until filter can be Unix timestamps, date formatted timestamps, or Go duration strings (e.g. 10m, 1h30m) computed relative to the daemon machine’s time.
```
Comment (ashley-cui) [inline docs/source/markdown/podman-system-prune.1.md:32]: Same

File: pkg/domain/infra/abi/system.go [hunk 4]
```diff
@@ -179,7 +180,16 @@ func (ic *ContainerEngine) SystemPrune(ctx context.Context, options entities.Sys
...
-		containerPruneReport, err := ic.pruneContainersHelper(nil)
+		containerPruneOptions := entities.ContainerPruneOptions{}
+		for _, f := range options.Filter {
+			t := strings.SplitN(f, "=", 2)
+			containerPruneOptions.Filters = make(url.Values)
+			if len(t) < 2 {
+				return nil, errors.Errorf("filter input must be in the form of filter=value: %s is invalid", f)
+			}
+			containerPruneOptions.Filters.Add(t[0], t[1])
+		}
```
Comment (bblenard) [inline pkg/domain/infra/abi/system.go:191]: When I was messing with this feature I noticed that this pattern is repeated in some form or another

```
cmd/podman/containers/prune.go:69:		t := strings.SplitN(f, "=", 2)
cmd/podman/containers/ps.go:189:		split := strings.SplitN(f, "=", 2)
cmd/podman/networks/list.go:70:		split := strings.SplitN(f, "=", 2)
cmd/podman/pods/ps.go:85:			split := strings.SplitN(f, "=", 2)
cmd/podman/volumes/list.go:76:		filterSplit := strings.SplitN(f, "=", 2)
libpod/image/prune.go:116:		filterSplit := strings.SplitN(f, "=", 2)
libpod/image/search.go:266:		arr := strings.SplitN(f, "=", 2)
```

Could this be abstracted into a helper function or is it not worth it? I thought maybe putting a `parseFilterArgumentsIntoMap` function into `libpod/util.go` or `cmd/podman/utils/utils.go` but I wasn't sure.

**Comment by bblenard** [comment]
> @bblenard I opened issues on volume --filter and on podman network prune, if you want to take a look at those.

I would love to take a swing at it.

File: pkg/domain/infra/abi/system.go (hunk 4 above)
Comment (rhatdan) [reply pkg/domain/infra/abi/system.go:191]: SGTM.  I think getting them altogether would be good.

**Comment by rhatdan** [comment]
@bblenard Just grab it.

File: docs/source/markdown/podman-system-prune.1.md (hunk 3 above)
Comment (rhatdan) [reply docs/source/markdown/podman-system-prune.1.md]: Thanks @ashley-cui
Fixed.

Comment (TomSweeneyRedHat) [reply docs/source/markdown/podman-system-prune.1.md]: I think we need to convene a kangaroo court and fine @rhatdan $25 to be donated to a charity of his choice for inserting daemon there!

**Comment by TomSweeneyRedHat** [comment]
LGTM

File: docs/source/markdown/podman-system-prune.1.md (hunk 3 above)
Comment (rhatdan) [reply docs/source/markdown/podman-system-prune.1.md]: More of a cut and paste, and missed removing it.

**Comment by svdHero** [comment]
@bblenard Thank you so much for your work. :+1: 

@rhatdan How can I test this most easily? So far, I have installed podman using apt on my Ubuntu machine. How would I download and install the master branch?

**Comment by bblenard** [comment]
> @bblenard Thank you so much for your work. 
> 
> @rhatdan How can I test this most easily? So far, I have installed podman using apt on my Ubuntu machine. How would I download and install the master branch?

The way I did it was to follow: https://github.com/containers/podman/blob/master/CONTRIBUTING.md#contributing-to-podman-1 through to the [building binaries](https://github.com/containers/podman/blob/master/CONTRIBUTING.md#building-binaries-and-test-your-changes) section and then using the `make binaries` target. It looks like @rhatdan's changes were merged into master so assuming you have an up to date copy of the repo you should be able to run `bin/podman system prune --filter ...` to test. The installation from the ubuntu repos will probably not include these changes so you will have to build the binaries yourself I'm guessing.

**Comment by rhatdan** [comment]
Correct you will not see these changes in a distribution (Other then Fedora/Rawhide) until we release podman 3.0.

**Pull request merged**